consensus/core/src/dag_state.rs

```diff
@@ -542,6 +547,12 @@ impl DagState {
             );
         }
 
+        // Ensure commit timestamp cannot go backwards.
+        let leader_block = self.get_block(&commit.leader()).unwrap();
+        self.last_commit_timestamp_ms = self
+            .last_commit_timestamp_ms
+            .max(leader_block.timestamp_ms());
+
```

Comment:
It might enhance the argument of keeping the commit timestamp in the `Commit` struct, but ideally I would suggest avoid calculating again the logic of timestamp here, as we do it already on the `collect_sub_dag` method. Otherwise the timestamp could be provided as part of the `add_commit` method, but this is also not ideal.

Reply:
Having duplicated logic between DagState and Linearizer is definitely not great, but I don't think the situation much different between timestamp and the existing last committed round logic. We can revisit this later together with scores. It seems all of them can be passed in via CommittedSubDag.